### DISCUSSION TITLE
Plugin: Synchronize via XMPP (no server needed)

### DISCUSSION THREAD

**Discussion by sjehuda** [lwindolf/liferea#1343] [Ideas]
I want to write a plugin using slixmpp library which would allow to synchronize Liferea instances via XMPP.

As I think of it, this would be feasible when two Liferea instances are running at the same time and - of course - connected to the same JID (Jabber ID - XMPP Account).

How does it work?
1) Liferea A and Liferea B are running at the same time from two different machines.
2) ~Liferea A receives updates...~ (I will think of this later).
3) Liferea A gets input to mark items as read or flag them.
4) Then Liferea A contacts Liferea B and tells it it's current status and changes.

* Both - Liferea A and Liferea B - are connected to the same JID, and each has a different resource `liferea@lzone.de/a` and `liferea@lzone.de/b`.

@lwindolf may I contact you by email for help?

I need to know how to get the list of the feeds and items, including state (unread, read, flag etc.).

If it is not possible, I will read data directly from the SQLite database.

@horstjens would you like to help test this plugin when it is ready?

**Comment by sjehuda** [comment]
I need help please.

1) I want to print the list of feeds.
2) I want to modify the list.

I can take care for all the rest of the code with slixmpp.

**Reply by lwindolf** [reply]
Feel free to email me with questions.

In Python try `Liferea.FeedList.get_root()` to access the feed list tree. Right now it will give you a C-type nodePtr I'm not 100% sure if the contents are usable for your purpose. Simple fields like id/url/name... should work.

For modifying you would use `Liferea.FeedList.add_folder()` and `Liferea.FeedList.add_subscription()`.

For available methods please check out `src/Liferea-3.0.gir`

**Reply by sjehuda** [reply]
I just sent you an email message.

Pardon for not checking this thread before sending an email.

I will try it and I will inform you of any progress made.

Thank you Lars!

**Reply by sjehuda** [reply]
> For available methods please check out `src/Liferea-3.0.gir`

Where is it please? I did not find it. Please provide a URL.

> `Liferea.FeedList.get_root()`

I get this error: `AttributeError: type object 'FeedList' has no attribute 'get_root'`

I migh be doing it incorrectly.

This is the code (from bold unread plugin)
```
from gi.repository import GObject, Liferea, Pango
import slixmpp

def cell_func_bold_unread(column, cell, model, iterator, data):
    unread_count = model.get(iterator, 3)
    cell.set_property("weight", Pango.Weight.BOLD if (unread_count[0] > 0) else Pango.Weight.NORMAL)

class SyncFeedXmppPlugin(GObject.Object, Liferea.ShellActivatable):
    __gtype_name__ = 'NrBoldUnreadPlugin'
    treeview = None
    ticol = None
    layout = None
    shell = GObject.property(type=Liferea.Shell)

    def do_activate(self):
        Liferea.FeedList.get_root()
        print(Liferea.FeedList.get_root())
        
        # Traceback (most recent call last):
        #   File "/home/admin/.local/share/liferea/plugins/feed-sybc/feed-sync.py", line 35, in do_activate
        #     Liferea.FeedList.get_root()
        #     ^^^^^^^^^^^^^^^^^^^^^^^^^
        # AttributeError: type object 'FeedList' has no attribute 'get_root'

        
        self.itemlist = self.shell.lookup("itemlist")
        print(self.itemlist)
        breakpoint()
        self.treeview = self.shell.lookup("feedlist")
        self.ticol = self.treeview.get_column(0)
        area = self.ticol.get_property("cell-area")
        self.layout = area.get_cells()
        self.ticol.set_cell_data_func(self.layout[1], cell_func_bold_unread)
        self.treeview.queue_draw()

    def do_deactivate(self):
        self.ticol.set_cell_data_func(self.layout[1], None)
        self.layout[1].set_property("weight", Pango.Weight.NORMAL)
        self.treeview.queue_draw()
```

**Reply by lwindolf** [reply]
`Liferea-3.0.gir` is created when you compile. It will be created in your source checkout under `src/`.

As for `FeedList.get_root()` you are right, you cannot use it as of now. The reason being that the node type is not implemented as a GObject. This actually means the a plugin modifying the feed list is not possible.

**Reply by sjehuda** [reply]
Is it now possible to modify the list of feeds?

**Reply by lwindolf** [reply]
Sadly not yet

**Reply by sjehuda** [reply]
Fine. No problems. I am currently still engaged with other projects.

I will ask again, later.

Thank you.
